### PR TITLE
Validate docs dir before writing custom js

### LINKED ISSUE #5568: Validate docs dir before writing custom js
If the user puts an invalid docs dir, we fail with a generic error, we should raise a more specific one.

https://github.com/rtfd/readthedocs.org/blob/44951fbf8fe1fe12fa9f99f3075f87bd77188525/readthedocs/doc_builder/backends/mkdocs.py#L175-L176

### PR DESCRIPTION
**Pull request by saadmk11**
closes #5568

### CODE REVIEW

File: readthedocs/doc_builder/base.py [hunk 1]
```diff
@@ -84,6 +86,13 @@ def clean(self, **__):
     def docs_dir(self, docs_dir=None, **__):
         """Handle creating a custom docs_dir if it doesn't exist."""
         checkout_path = self.project.checkout_path(self.version.slug)
+        if docs_dir:
+            possible_path = os.path.join(checkout_path, docs_dir)
+            # if user puts an invalid `docs_dir` path raise Exception
+            if not os.path.exists(possible_path):
+                raise MkDocsYAMLParseError(
+                    MkDocsYAMLParseError.INVALID_DOCS_DIR_PATH,
+                )
```
Comment (stsewd) [inline readthedocs/doc_builder/base.py]: We shouldn't put special cases in the base class, the base class should know nothing about mkdocs.

Comment (saadmk11) [reply readthedocs/doc_builder/base.py]: Got it :)

Comment (saadmk11) [reply readthedocs/doc_builder/base.py]: @stsewd Updated the PR

File: readthedocs/doc_builder/backends/mkdocs.py [hunk 2]
```diff
@@ -118,7 +118,18 @@ def append_conf(self, **__):
                 MkDocsYAMLParseError.INVALID_DOCS_DIR_CONFIG,
             )
 
+        # if user puts an invalid `docs_dir` path raise an Exception
+        if user_docs_dir:
+            checkout_path = self.project.checkout_path(self.version.slug)
+            possible_path = os.path.join(checkout_path, user_docs_dir)
```
Comment (stsewd) [inline readthedocs/doc_builder/backends/mkdocs.py]: We actually need to check if this path doesn't exists https://github.com/rtfd/readthedocs.org/blob/4fea14e86fe0aabffd73352402ddf8dbba39ff40/readthedocs/doc_builder/backends/mkdocs.py#L159-L162

File: readthedocs/rtd_tests/tests/test_doc_builder.py [hunk 3]
```diff
@@ -488,6 +488,42 @@ def test_write_js_data_docs_dir(self, checkout_path, run, generate_rtd_data):
...
+            build_env=self.build_env,
+            config=None,
+        )
+        self.searchbuilder = MkdocsHTML(
+            build_env=self.build_env,
+            python_env=python_env,
+        )
+        with self.assertRaises(MkDocsYAMLParseError):
+            self.searchbuilder.append_conf()
+            generate_rtd_data.assert_called_with(
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_builder.py]: I guess we don't need this assert

File: readthedocs/doc_builder/backends/mkdocs.py (hunk 2 above)
Comment (saadmk11) [reply readthedocs/doc_builder/backends/mkdocs.py]: Oh! you meant `docs_path` I thought you meant the `docs_dir` which user adds in the yaml. okay I'll Update the PR.

Comment (stsewd) [reply readthedocs/doc_builder/backends/mkdocs.py]: Just to be clear, it is the one the users put in the mkdocs.yml file

Comment (stsewd) [reply readthedocs/doc_builder/backends/mkdocs.py]: But the docs dir is relative to the mkdocs.yml file, which can be located in other places not jus in the root

File: readthedocs/rtd_tests/tests/test_doc_builder.py (hunk 3 above)
Comment (saadmk11) [reply readthedocs/rtd_tests/tests/test_doc_builder.py]: @stsewd then how should we check if invalid `docs_path` raises an error or not?

Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_doc_builder.py]: Yes, we only need that check. The second assert wouldn't happen anyway.

File: readthedocs/doc_builder/backends/mkdocs.py (hunk 2 above)
Comment (saadmk11) [reply readthedocs/doc_builder/backends/mkdocs.py]: @stsewd Thanks for the explanation. :+1:

File: readthedocs/rtd_tests/tests/test_doc_builder.py (hunk 3 above)
Comment (saadmk11) [reply readthedocs/rtd_tests/tests/test_doc_builder.py]: yes that's True. :+1:  updated the PR.

**Pull request merged**